redwood: Add is_valid_public_key function

legoktm opened the pull request:
## Status

Ready for review; ~~but will fail CI until #6892 lands (this will be rebased on top of that prior to merge)~~

## Description of Changes

We have multiple cases in Python code where we want to validate that a
string is valid PGP public key that Sequoia will accept. Exporting a
`is_valid_public_key` function will help make checks more robust and
simplify some tests.
    
The validation code on the Rust side is also a bit duplicated and
currently manually kept in sync with comments, so let's take the
opportunity to unify it into one file and de-duplicated with a macro
(hopefully not too magical).
    
Fixes #6947.

## Testing

* [ ] Visual review
* [ ] CI passes

## Deployment

Any special considerations for deployment? no

## Checklist

- [x] Linting (`make lint`) and tests (`make test`) pass in the development container

legoktm committed:
redwood: Add is_valid_public_key function

We have multiple cases in Python code where we want to validate that a
string is valid PGP public key that Sequoia will accept. Exporting a
`is_valid_public_key` function will help make checks more robust and
simplify some tests.

The validation code on the Rust side is also a bit duplicated and
currently manually kept in sync with comments, so let's take the
opportunity to unify it into one file and de-duplicated with a macro
(hopefully not too magical).

Tests are provided that verify a good public key passes, a bad SHA-1
key is rejected and a secret key is also rejected.

Fixes #6947.

legoktm committed:
Use redwood.is_valid_public_key() in tests

This is a more comprehensive check than just verifying some variant of
the string "BEGIN PGP PUBLIC KEY BLOCK" is present.

legoktm commented on redwood/src/keys.rs:
```diff
@@ -0,0 +1,55 @@
+/// We want to use the same iterators on public and secret keys but it's not
+/// really possible to do it in a function because of type differences so we
+/// use a macro instead.
+macro_rules! filter_keys {
```
I spent a while trying to write a generic function that applied these filters over public key parts and secret key parts (distinct types) and 1) it was super complicated and 2) I couldn't actually get it to work/compile.

So I went with a macro, but I'm also okay saying macros are too magical for our collective current level of Rust and to just keep duplicating for now.

rocodes replied:
(I vote leave it in - it is clear/readable, basically by the book, and less error-prone than manual duplication of the same logic in different places)

legoktm commented:
Marking as ready for review; CI will fail because SD tests are generating 1024-bit keys until #6892 lands. But at least the code can be reviewed ahead of time.

rocodes commented on redwood/src/lib.rs:
```diff
@@ -94,6 +98,19 @@ pub fn generate_source_key_pair(
 
+#[pyfunction]
+pub fn is_valid_public_key(input: &str) -> Result<String> {
+    let policy = &StandardPolicy::new();
```
(Very minor point but) if the code becomes more complex or we ever stop using StandardPolicy, we may want to create one policy instance and pass it to `encrypt` and `is_valid_pubkey` to ensure there's no developer error and we're always testing and encrypting against the same policy.

legoktm replied:
That's a good point, and since you mentioned it, I looked, `StandardPolicy::new()` is const, so we can in fact already do that! Commit incoming...

legoktm committed:
redwood: Use a single StandardPolicy instance

StandardPolicy::new() is const, so we can create one instance of
the policy and use it everywhere, ensuring that we don't accidentally
use different policies.

rocodes reviewed (dismissed):
LGTM, and very polite way of making improvements to my attempt at enforcing the key policy :see_no_evil:  

Waiting til https://github.com/freedomofpress/securedrop/pull/6928 lands but otherwise :rocket:

cfm reviewed (dismissed):
This is just a drive-by review after #6892 and before #6946.  Very nice addition and refactoring!  I'll defer to you two for the final rebase and merge.
